Source/Urho3D/UI/UIComponent.h

```diff
@@ -61,29 +65,41 @@ class URHO3D_API UIComponent : public Component
     /// Model that texture will be applied to.
     SharedPtr<StaticModel> model_;
     /// UIElement to be rendered into texture.
-    SharedPtr<UIElement> rootElement_;
-    /// UI rendering batches.
-    PODVector<UIBatch> batches_;
-    /// UI rendering vertex data.
-    PODVector<float> vertexData_;
-    /// UI vertex buffer.
-    SharedPtr<VertexBuffer> vertexBuffer_;
-    /// UI rendering batches for debug draw.
-    PODVector<UIBatch> debugDrawBatches_;
-    /// UI rendering vertex data for debug draw.
-    PODVector<float> debugVertexData_;
-    /// UI debug geometry vertex buffer.
-    SharedPtr<VertexBuffer> debugVertexBuffer_;
+    SharedPtr<UIElement3D> rootElement_;
     /// Is StaticModel component created by this component.
     bool isStaticModelOwned_;
 
+    /// Handle component being added to Node or removed from it.
     virtual void OnNodeSet(Node* node) override;
     /// Handle resizing of element. Setting size of element will automatically resize texture. UIElement size matches size of texture.
     void OnElementResized(StringHash eventType, VariantMap& args);
-    /// Convert screen position to position on UIElement.
-    bool ScreenToUIPosition(IntVector2 screenPos, IntVector2& result);
+};
+
+class URHO3D_API UIElement3D : public UIElement
+{
+    URHO3D_OBJECT(UIElement3D, UIElement);
+public:
+    /// Construct.
+    UIElement3D(Context* context);
+    /// Destruct.
+    virtual ~UIElement3D() override = default;
+    /// Register object factory.
+    static void RegisterObject(Context* context);
 
-    friend class UI;
+    /// Set UIComponent which is using this element as root element.
+    void SetUIComponent(UIComponent* component);
```

Comment:
Looks like hierarchy flaw because `UIElement3D` has bidirectional connection with `UIComponent`. What's the point of splitting one class into two if they both still _require_ each other to work?

Reply:
`UIElement3D` acts as a root UI element for layout rendered to a texture. It handles screen -> element mouse translation. `UIComponent` handles UI texture rendering on to an object. They do different things. Problem is that in order to do mouse coordinates translation we need to know where object is rendered. So.. Whats the alternative? UI subsystem needs to query translated coordinates and this is by far cleanest way i could think of.